### ISSUE TITLE
node.js module (@bugsnag/node) does not allow to be bundled

### ISSUE THREAD

**Issue by kwonoj** [bugsnag/bugsnag-js#441]
### Reproducible code:

https://github.com/kwonoj/bugsnag-node-webpack

`npm install && npm start`

will raise exception around module resolution:

```
Error: Cannot find module 'error-stack-parser'
    at o (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:292)
    at /Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:449
    at Object.20../lib/es-utils (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:1229:24)
    at o (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:398)
    at /Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:449
    at Object.3../breadcrumb (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:164:21)
    at o (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:398)
    at /Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:449
    at Object.24../config (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:1638:14)
    at o (/Users/ojkwon/github/bugsnag-node-webpack/dist/index.bundle.js:109:398)
```

This looks like distributed module `@bugsnag/node` is already bundled somehow, cannot resolve module by consumer application's bundler configuration.

### Motivation

Yes, typically node.js side module is not intended to be bundled. While reproducible code is provided as targeting `node` in example, actual usecases for our application is targeting `electron-main`, which inherits node.js nature blocks to create application bundle including bugsnag but have to specify as external. We would like to remove externals as it defeats our purpose of bundling.

### Possible approaches

If it's possible, I'd like to suggest to publish module with non-bundled code. This can be non-breaking changes if publishing esm module for those purpose (using `package.json`'s `module` field) while preserving existing main field with current bundled modules.

**Comment by brimtown** [comment]
Running into the same issue with `@bugsnag/js` and `@bugsnag/react` in Gatsby, specifically when attempting to use `bugsnagClient.notify` manually from code that is run through both Node and the browser.

Really looking forward to these packages being usable, as it's currently almost impossible to do manual `notify` calls in Gatsby!

**Comment by axelboc** [comment]
Same problem here doing SSR with `@bugsnag/js` and `@bugsnag/plugin-angular`

**Comment by bengourley** [comment]
Hi all, thanks for the report and especially the steps to reproduce. This was quite a tricky problem to solve but we got there in the end. I just wanted to leave a bunch of information here, as this seems like an issue that other people compiling for node with webpack will be running in to!

### Background

The result of all the `require()` calls in the node notifer is a dependency tree that looks like this:

```
@bugsnag/node
├─┬─ @bugsnag/core/client
│ └── iserror
├─┬─ @bugsnag/delivery-node
│ ├── @bugsnag/core/lib/json-payload
│ ├── @bugsnag/core/lib/es-utils
│ └── request
├─── @bugsnag/core/config
├─┬─ @bugsnag/plugin-node-surrounding-code
│ ├── byline
│ └── pump
├─┬─ @bugsnag/plugin-node-in-project
│ ├── @bugsnag/core/lib/es-utils
│ └── @bugsnag/core/lib/path-normalizer
... etc.
```

This is quite a busy dependency tree. The modularity that this brings has a benefit for development, and clean organisation of the project, however

- `require()` calls in node are synchronous file operations that happen frequently at startup. Too many of these can have a performance hit.
- All of the @bugsnag prefixed modules would take a long time to resolve and install, even though we control all of them and version them together in the same monorepo.

For these reasons, we bundle all of our own modules into a single dist file using browserify’s node mode. This means that all of the `require()` calls get analysed at bundle time and combined into the same file.

We exclude the external modules – those that we don’t control – using the `exclude` property of browserify (you can see some of these in the dependencies listed above):

https://github.com/bugsnag/bugsnag-js/blob/80bc6179b96ff1eac5ec711284009f9a25b34ad4/packages/node/package.json#L22

This means that when the bundle runs in Node, `require()` calls for these modules fall through to the native `require` and load them from `node_modules`.

The resulting dependency tree is something like this:

```
@bugsnag/node
├─── iserror
├─── stack-generator
├─── error-stack-parser
├─── request
├─── pump
└─── byline
```

Which is much more palatable at install time, and also quicker when the application boots. This is fine for the typical use case of node and `node_modules`.

### The problem

The problem arises when one attempts to bundle their Node app into a single file using Webpack. The reason for doing this for an application is perfectly reasonable, just lesson common.

When webpack bundles the prebuilt `@bugsnag/node`, its static analysis process does not detect the external `require()` calls in the code. It sticks a var require; in its own bundle just in front of `@bugsnag/node`, which means code within `@bugsnag/node` can’t access the built-in Node `require()` function to load any of its external dependencies from the disk.

### Potential (unworkable) solutions

- As suggested by OP, we could distribute an unbundled version of the code alongside the bundled version using the module key in package.json. This would probably help, but there are two problems:
  - we’d need to rewrite the module using es module `import`/`export` module syntax either by hand or by machine
  - we’d need to list all of the `@bugsnag/…` modules as dependencies in `package.json` meaning that people using the standard node version would have to install them all too, even though they have them all bundled.
- We wrap the external modules up in the bundle as well. The problems with this are
  - applications that share any dependencies with `@bugsnag/node` will not get to share a copy in `node_modules` – there will be two copies (potentially of _exactly_ the same version).
  - the resulting bundle file size will get rather huge, because we use the behemoth request (whose file structure on-disk is a shade under 5MB 😱).

### Actual solution

See #450 for more detail. In summary, using a browserify plugin which replaces the `require`/`module.exports` calls with simple assignments, means that webpack can statically analyse the file correctly.

**Comment by bengourley** [comment]
#450 shipped in [v5.1.0](https://github.com/bugsnag/bugsnag-js/releases/tag/v5.1.0).

You should now be able to consume @bugsnag/js` in a webpack bundle targeting node. Please let me know how you get on!
